core/src/openid.cpp

```diff
@@ -102,12 +102,100 @@ namespace irods::http::openid
 
-		// Validate access token
+		// Check that the access token is active.
+		// The exact meaninng of an active state may vary between OpenID Providers,
+		// so we should assert the other attributes when available.
 		if (!json_res.at("active").get<bool>()) {
 			logging::warn("{}: Access token is invalid or expired.", __func__);
 			return std::nullopt;
 		}
 
+		// The rest of the members are optional, but validate the ones that do exist.
+		// TODO: Verify that this is actually how this works. I don't think we can check this 'client_id'.
+		if (auto client_id_iter{json_res.find("client_id")}; client_id_iter != std::end(json_res)) {
+			if (client_id_iter->get_ref<const std::string&>() != irods::http::globals::oidc_configuration().at("client_id")) {
+				return std::nullopt;
+			}
+		}
+
+		// While not required, ensure the token type is something we can use.
+		// TODO: Consider removing this part. Knowing the 'token_type' after the server has already
+		// processed the token doesn't do anything for us.
+		if (auto token_type_iter{json_res.find("token_type")}; token_type_iter != std::end(json_res)) {
+
+			// There are only 'bearer' tokens and 'mac' tokens defined
+			if (token_type_iter->get_ref<const std::string&>() != "bearer") {
+				return std::nullopt;
+			}
+		}
+
+		// We must not accept expired tokens.
+		if (auto exp_iter{json_res.find("exp")}; exp_iter != std::end(json_res)) {
+			// Get current time
+			std::chrono::time_point<std::chrono::system_clock> now{};
+
+			// Convert to appropriate representation
+			std::chrono::seconds exp_seconds{exp_iter->get<int>()};
+
+			// Setup expire time
+			std::chrono::time_point<std::chrono::system_clock> expire_time{exp_seconds};
+
+			// Reject expired tokens
+			if (now > expire_time) {
+				return std::nullopt;
+			}
+		}
+
+		// While not strictly required, we shouldn't accept tokens from the future.
+		if (auto iat_iter{json_res.find("iat")}; iat_iter != std::end(json_res)) {
+			// Get current time
+			std::chrono::time_point<std::chrono::system_clock> now{};
+
+			// Convert to appropriate representation
+			std::chrono::seconds iat_seconds{iat_iter->get<int>()};
+
+			// Setup expire time
+			std::chrono::time_point<std::chrono::system_clock> issued_time{iat_seconds};
+
+			// Reject tokens from the future
+			if (now < issued_time) {
+				return std::nullopt;
+			}
+		}
+
+		// Token must not be used early if given 'not before' time.
+		if (auto nbf_iter{json_res.find("nbf")}; nbf_iter != std::end(json_res)) {
+			// Get current time
+			std::chrono::time_point<std::chrono::system_clock> now{};
+
+			// Convert to appropriate representation
+			std::chrono::seconds nbf_seconds{nbf_iter->get<int>()};
+
+			// Setup expire time
+			std::chrono::time_point<std::chrono::system_clock> not_before_time{nbf_seconds};
+
+			// Reject tokens that are used too early
+			if (now < not_before_time) {
+				return std::nullopt;
+			}
+		}
```

Comment:
Processing of `exp`, `iat`, and `nbf` should not be required. The OpenID Provider should validate the token for us (e.g. ensuring token is not used before `nbf`) and represent the validity of a token using the `active` member.

Reply:
Removed.